xray/commands/audit/npm.go

```diff
@@ -0,0 +1,146 @@
+)
+
+type XrAuditNpmCommand struct {
+	serverDetails    *config.ServerDetails
+	workingDirectory string
+	arguments        []string
+	typeRestriction  npm.TypeRestriction
+}
+
+func (auditCmd *XrAuditNpmCommand) SetWorkingDirectory(dir string) *XrAuditNpmCommand {
+	auditCmd.workingDirectory = dir
+	return auditCmd
+}
+
+func (auditCmd *XrAuditNpmCommand) SetArguments(args []string) *XrAuditNpmCommand {
+	auditCmd.arguments = args
+	return auditCmd
+}
+
+func (auditCmd *XrAuditNpmCommand) SetNpmTypeRestriction(typeRestriction npm.TypeRestriction) *XrAuditNpmCommand {
+	auditCmd.typeRestriction = typeRestriction
+	return auditCmd
+}
+
+func (auditCmd *XrAuditNpmCommand) SetServerDetails(server *config.ServerDetails) *XrAuditNpmCommand {
+	auditCmd.serverDetails = server
+	return auditCmd
+}
+
+func (auditCmd *XrAuditNpmCommand) ServerDetails() (*config.ServerDetails, error) {
+	return auditCmd.serverDetails, nil
+}
+
+func NewXrAuditNpmCommand() *XrAuditNpmCommand {
+	return &XrAuditNpmCommand{}
+}
+
+func (na XrAuditNpmCommand) Run() (err error) {
+	nca := npm.NewNpmCommandArgs("")
+	nca.SetTypeRestriction(na.typeRestriction)
+
+	currentDir, err := commandsutils.GetWorkingDirectory()
+	if err != nil {
+		return err
+	}
+	if na.workingDirectory == "" {
+		na.workingDirectory = currentDir
+	} else {
+		err = os.Chdir(na.workingDirectory)
+		if err != nil {
+			return err
+		}
+		defer func() {
+			err = os.Chdir(currentDir)
+		}()
+	}
+	log.Debug("Working directory set to:", na.workingDirectory)
+
+	packageInfo, err := commandsutils.ReadPackageInfoFromPackageJson(na.workingDirectory)
+	if err != nil {
+		return err
+	}
+	nca.SetPackageInfo(packageInfo)
+	err = nca.SetNpmExecutable()
+	if err != nil {
+		return err
+	}
+	// Calculate npm dependencies
+	err = nca.SetDependenciesList()
+	if err != nil {
+		return err
+	}
+	// Parse the dependencies into an Xray dependency tree format
+	npmGraph := parseNpmDependenciesList(nca.GetDependenciesList(), packageInfo)
+	xrayManager, err := commands.CreateXrayServiceManager(na.serverDetails)
+	if err != nil {
+		return err
+	}
+	params := services.NewXrayGraphScanParams()
+	params.Graph = npmGraph
+	scanId, err := xrayManager.ScanGraph(params)
+	if err != nil {
+		return err
+	}
+
+	scanResults, err := xrayManager.GetScanGraphResults(scanId)
+	if err != nil {
+		return err
+	}
+	return printTable(scanResults)
+
+}
+
+func printTable(res *services.ScanResponse) error {
+	jsonOut, err := json.Marshal(res)
+	print(string(jsonOut))
+	return err
+}
+
+func parseNpmDependenciesList(dependencies map[string]*npm.Dependency, packageInfo *commandsutils.PackageInfo) (xrDependencyTree *services.GraphNode) {
+	treeMap := make(map[string][]string)
+	for dependencyId, dependency := range dependencies {
+		dependencyId = NpmPackageTypeIdentifier + dependencyId
+		// Because we are dealing with a
+		father := NpmPackageTypeIdentifier + dependency.GetPathToRoot()[0][0]
+		if sons, ok := treeMap[father]; ok {
```

Comment:
Let's keep the convention of parent-child
```suggestion
		parent := NpmPackageTypeIdentifier + dependency.GetPathToRoot()[0][0]
		if children, ok := treeMap[father]; ok {
```